nntrainer/src/neuralnet.cpp

```diff
@@ -442,17 +443,16 @@ int NeuralNetwork::initLossLayer() {
       ml_loge("Error: Cross Entropy not supported without softmax or sigmoid.");
       return ML_ERROR_NOT_SUPPORTED;
     }
```

Comment:
The case has moved to here since network can evaluate if the last activation layer(if any) can be integrated into loss layer or just leave the way it is.

Reply:
How is `COST_MSR` handled here?

Reply:
`COST_MSR` does not need to be handled here. Activation layer is added when `Network::init()` as long as user put activation props for the last layer .

The statement that `activation layer is last layer before adding loss layer` is not guaranteed though, it can be seen as user's fault. Do you think we need to warn the user if last layer is not activation layer before adding loss layer?

Reply:
I meant that in case `fc_layer` is the last layer, gradient is handled differently for `COST_MSR` and `COST_ENTROPY`. I found that missing here.

Reply:
This patch operates in this part as follows:

1. Add activation layer if `fc_layer` has activation prop.
2. in case of `COST_MSR`, just add `COST_MSR` loss layer
3. in case of `COST_ENTROPY` : 
   3-1. if last layer is not activation layer, throw error
   3-2. if last layer is sigmoid or softmax, **pop_back** that activation layer, and add `COST_ENTROPY_*` loss_layer.
   
```cpp
  if(!last_layer){
    djdb = derivative.multiply(hidden.apply(activation_prime));
 } switch (cost) {	
    case COST_MSR:	
      // this case does not need handling since activation layer(should have) added after last fc_layer (in #2).
      if (activation_type == ACT_SOFTMAX) {	
        y = hidden.apply(softmax);	
        djdb = derivative.multiply(hidden.apply(softmaxPrime));	
      } else {	
        djdb = derivative.multiply(hidden.apply(activation_prime));	
      }	
     
      break;	

    // since activation layer is removed before adding loss layer (#3-2) this case can be removed as well.
    case COST_ENTROPY_SIGMOID:	
      /** intended */	
    case COST_ENTROPY_SOFTMAX:	
      djdb = derivative;	
      break;	
    
    // shouldn't reach this part.
    case COST_ENTROPY:	
      throw std::runtime_error(	
        "Error: Cross Entropy not supported without softmax or sigmoid.");	
    case COST_UNKNOWN:	
      /** Intended */
```

revised

```cpp

 if (!last_layer) {
   djdb = derivative //.multiply(hidden.apply(activation) is done in activation layer 
 } else {
   switch (cost) {
    case COST_MSR:
      djdb = derivative; //.multiply(hidden.apply(...)) is done in activation layer
      break;
    case COST_ENTROPY_SIGMOID:
    case COST_ENTORPY_SOFTMAX:
      djdb = derivative;
      break;
    case OTHERS:
      //omitted
  } 
 }
```

So there is no need for a case handling. If my explanation is out of tune. I guess we can discuss this matter offline.

Out of topic though, I think there is a bug with this implementation.
Current implementation is `djdb = derivatives` so extra variable is not used. However, I think `djdb = derivatives.sum_by_batch()`. isn't it?

Reply:
Further, applying activation prime for softmax and others is also different for COST_MSR.

Reply:
> Further, applying activation prime for softmax and others is also different for COST_MSR.

It was just because before this patch `activation_prime` was `std::function<float (float)>` which cannot handle softmax.

Activation_layer.cpp handles them all together(checkout `activation_layer.cpp:76~122`)

Reply:
Thanks for the explanation.
I understand how the `COST_MSR` is handled now. I think there is just 1 missing point about the softmax activation type.
For generic activation layers - 
```
djdb = derivative.multiply(hidden.apply(activation_prime));	
```
where hidden is the output of the fc/conv layer and activation is not yet applied.

However, for softmax
```
y = derivative.apply(softmax);
djdb = derivative.multiply(y.apply(activation_prime));
```
Attached an old version of this file https://github.com/nnstreamer/nntrainer/blob/2a650512813db6ce3bba828b5790066fbc655f14/nntrainer/src/fc_layer.cpp#L264-L267

I have added a PR #223 to reflect it back to the original version.
Please let me know if I understood it wrong.